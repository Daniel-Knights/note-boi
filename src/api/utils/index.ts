export * from './decorators';
export * from './error';
export * from './http';
