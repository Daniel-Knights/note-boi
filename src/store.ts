import { reactive } from 'vue';
import { invoke } from '@tauri-apps/api/tauri';
import { v4 as uuidv4 } from 'uuid';

import { isEmptyNote } from './utils';

export class Note {
  readonly id = uuidv4();
  body = {
    delta: '',
    text: '',
  };
  timestamp = Date.now();
}

interface State {
  notes: Note[];
  selectedNote: Note;
}

export const state = reactive<State>({
  notes: [],
  selectedNote: new Note(),
});

/** Sorts notes in descending order by timestamp. */
function sortNotes() {
  state.notes.sort((a, b) => b.timestamp - a.timestamp);
}

/** Finds note index within {@link state.notes}. */
function findNoteIndex(id: string) {
  return state.notes.findIndex((nt) => nt.id === id);
}

/** Finds note within {@link state.notes}. */
export function findNote(id: string): Note | undefined {
  return state.notes.find((nt) => nt.id === id);
}

/** Deletes {@link state.selectedNote} when both `title` and `body` are empty. */
function clearEmptyNote(): void {
  const isValidClear = state.notes.length > 1;

  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote) return;

  if (isValidClear && isEmptyNote(foundNote)) {
    deleteNote(state.selectedNote.id);
  }
}

const selectEvent = new CustomEvent('note-select');

/**
 * Looks for note with given `id` in {@link state.notes},
 * and sets it to {@link state.selectedNote}.
 */
export function selectNote(id: string): void {
  if (state.selectedNote.id === id) return;

  clearEmptyNote();

  const foundNote = findNote(id);
  if (foundNote) state.selectedNote = { ...foundNote };

  document.dispatchEvent(selectEvent);
}

/** Fetches all notes and updates {@link state}. */
export async function getAllNotes(): Promise<void> {
  const fetchedNotes = await invoke<Note[]>('get_all_notes').catch(console.error);

  const hasNotes = fetchedNotes && fetchedNotes.length > 0;
  if (!hasNotes) return newNote();

  state.notes = fetchedNotes;

  sortNotes();

  if (isEmptyNote(fetchedNotes[0])) {
    state.notes[0].timestamp = Date.now();
  }

  state.selectedNote = { ...state.notes[0] };

  document.dispatchEvent(selectEvent);
}

/** Deletes note with the given `id`. */
export function deleteNote(id: string): void {
  state.notes.splice(findNoteIndex(id), 1);
  state.selectedNote = { ...state.notes[0] };

  sortNotes();

  invoke('delete_note', { id }).catch(console.error);
}

/** Creates an empty note. */
export function newNote(): void {
  const foundNote = findNote(state.selectedNote.id);

  // Only update timestamp if selected note is empty
  if (foundNote && isEmptyNote(foundNote)) {
    state.selectedNote.timestamp = Date.now();
    return;
  }

  const freshNote = new Note();

  state.notes.unshift(freshNote);
  state.selectedNote = { ...freshNote };

  invoke('new_note', { note: { ...freshNote } }).catch(console.error);

  document.dispatchEvent(selectEvent);
}

/** Edits note body on Quill `text-change`. */
export function editBody(delta: string, text: string): void {
  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote || text === foundNote.body.text) return;

  const timestamp = Date.now();
  foundNote.timestamp = timestamp;
  state.selectedNote.timestamp = timestamp;

  foundNote.body = { delta, text };

  sortNotes();

  invoke('edit_note', { note: { ...foundNote } }).catch(console.error);
}
