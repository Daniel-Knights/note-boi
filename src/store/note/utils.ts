import { Note } from '../../classes';
import { isEmptyNote } from '../../utils';

import { deleteNote } from './operations';
import { noteState } from './state';

/** Sorts notes in descending order by timestamp. */
export const sortNotesFn = (a: Note, b: Note): number => b.timestamp - a.timestamp;

/** Sorts {@link noteState.notes} in descending order by timestamp. */
export const sortStateNotes = (): Note[] => noteState.notes.sort(sortNotesFn);

/** Finds note index within {@link noteState.notes}. */
export function findNoteIndex(uuid?: string): number {
  return noteState.notes.findIndex((nt) => nt.uuid === uuid);
}

/** Finds note within {@link noteState.notes}. */
export function findNote(uuid?: string): Note | undefined {
  return noteState.notes.find((nt) => nt.uuid === uuid);
}

/**
 * Returns true if note is either {@link noteState.selectedNote}
 * or within {@link noteState.extraSelectedNotes}.
 */
export function isSelectedNote(note: Note): boolean {
  return (
    note.uuid === noteState.selectedNote.uuid ||
    noteState.extraSelectedNotes.some((nt) => nt?.uuid === note.uuid)
  );
}

/** Deletes {@link noteState.selectedNote} when note is empty. */
export function clearEmptyNote(): void {
  const isValidClear = noteState.notes.length > 1;
  if (!isValidClear) return;

  const foundNote = findNote(noteState.selectedNote.uuid);
  if (!foundNote) return;

  if (isEmptyNote(foundNote)) {
    deleteNote(noteState.selectedNote.uuid);
  }
}
