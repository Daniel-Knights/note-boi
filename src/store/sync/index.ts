import { http } from '@tauri-apps/api';
import type { Response } from '@tauri-apps/api/http';
import { reactive } from 'vue';

import { STORAGE_KEYS } from '../../constant';
import { isDev } from '../../utils';
import { UnsyncedNoteIds, unsyncedNoteIds } from './note';

export enum ErrorType {
  None,
  Auth,
  Push,
  Pull,
  Logout,
}

export const syncState = reactive({
  username: localStorage.getItem(STORAGE_KEYS.USERNAME) || '',
  password: '',
  token: localStorage.getItem(STORAGE_KEYS.TOKEN) || '',
  isLoading: false,
  isLogin: true, // For switching login/signup form
  error: {
    type: ErrorType.None,
    message: '',
  },
  unsyncedNoteIds: <UnsyncedNoteIds>{
    new: unsyncedNoteIds.new || '',
    edited: new Set<string>(unsyncedNoteIds.edited),
    deleted: new Set<string>(unsyncedNoteIds.deleted),
    get size() {
      return this.edited.size + this.deleted.size;
    },
    clear() {
      this.new = '';
      this.edited.clear();
      this.deleted.clear();
      localStorage.removeItem(STORAGE_KEYS.UNSYNCED);
    },
    add(ids): void {
      if (ids.new !== undefined) this.new = ids.new;

      ids.edited?.forEach((id) => this.edited.add(id));

      // Only track deleted notes if logged in
      ids.deleted?.forEach((id) => {
        this.deleted.add(id);

        if (this.edited.has(id)) {
          this.edited.delete(id);
        }
      });

      if (this.edited.has(this.new) || this.deleted.has(this.new)) {
        this.new = '';
      }

      localStorage.setItem(
        STORAGE_KEYS.UNSYNCED,
        JSON.stringify({
          new: this.new,
          edited: [...this.edited],
          deleted: [...this.deleted],
        })
      );
    },
  },
});

// Utils //

/** Parses an error response and returns a formatted message. */
export function parseErrorRes(res: Response<Record<string, unknown>>): string {
  const unknownErrorMessage = 'Unknown error, please try again';

  if (!res.data) return unknownErrorMessage;

  return typeof res.data.error === 'string' ? res.data.error : unknownErrorMessage;
}

/** Resets {@link state.error}. */
export function resetError(): void {
  syncState.error = { type: ErrorType.None, message: '' };
}

/** Wrapper for {@link http.fetch}. */
export function tauriFetch<T>(
  endpoint: string,
  method: 'POST' | 'PUT',
  payload?: Record<string, unknown>
): Promise<Response<T>> {
  const baseUrl = isDev()
    ? 'http://localhost:8000'
    : 'https://note-boi-server.herokuapp.com';

  return http.fetch<T>(`${baseUrl}/api${endpoint}`, {
    method,
    body: {
      type: 'Json',
      payload,
    },
  });
}

/** Catches hanging requests (e.g. due to server error). */
export function catchHang(err: unknown, type: ErrorType): void {
  console.error(err);

  syncState.isLoading = false;
  syncState.error = {
    type,
    message: 'Request failed',
  };
}

export { autoPush, push, pull, syncNotes, UnsyncedNoteIds } from './note';
export { clientSideLogout, login, signup, logout } from './auth';
export { deleteAccount } from './account';
