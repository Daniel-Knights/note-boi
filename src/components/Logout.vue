<template>
  <button v-if="state.token" @click="logout" id="logout" title="Logout">
    <LogoutIcon />
  </button>
</template>

<script lang="ts" setup>
import { logout, state } from '../store/sync';

import LogoutIcon from './svg/LogoutIcon.vue';
</script>

<style lang="scss" scoped></style>
