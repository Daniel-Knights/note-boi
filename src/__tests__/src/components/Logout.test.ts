import { mount } from '@vue/test-utils';

import * as s from '../../../store/sync';
import { clearMockApiResults, mockApi, mockDb } from '../../api';
import { awaitSyncLoad } from '../../utils';

import Logout from '../../../components/Logout.vue';

describe('Logout', () => {
  it('Mounts', async () => {
    const { calls, promises } = mockApi();
    const wrapper = mount(Logout);

    await Promise.all(promises);

    assert.isFalse(wrapper.isVisible());
    assert.strictEqual(calls.size, 0);
  });

  it('Logs out on click', async () => {
    const { calls } = mockApi({
      request: {
        resValue: {
          '/login': [{ notes: mockDb.encryptedNotes, token: 'token' }],
        },
      },
    });

    const wrapper = mount(Logout);
    assert.isFalse(wrapper.isVisible());

    s.syncState.username = 'd';
    s.syncState.password = '1';

    await s.login();

    assert.isTrue(wrapper.isVisible());

    clearMockApiResults({ calls });

    await wrapper.trigger('click');
    await awaitSyncLoad();

    assert.isEmpty(s.syncState.token);
    assert.strictEqual(calls.size, 2);
    assert.isTrue(calls.request.has('/logout'));
    assert.isTrue(calls.emits.has('logout'));
  });
});
