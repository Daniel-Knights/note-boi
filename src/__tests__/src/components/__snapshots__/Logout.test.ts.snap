// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Logout > Logs out on click > KeyStore.getKey() 1`] = `undefined`;

exports[`Logout > Logs out on click > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Logout > Logs out on click > Storage.get('THEME') 1`] = `null`;

exports[`Logout > Logs out on click > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Logout > Logs out on click > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Logout > Logs out on click > Storage.get('USERNAME') 1`] = `null`;

exports[`Logout > Logs out on click > allCalls 1`] = `
[
  [
    "request",
    {
      "calledWith": {
        "body": "{"username":"d","password":"1","notes":[],"deleted_note_ids":[]}",
        "headers": {
          "Content-Security-Policy": "default-src 'self'; connect-src http://localhost:8000;",
          "Content-Type": "application/json",
        },
        "method": "POST",
      },
      "name": "/auth/login",
    },
  ],
  [
    "invoke",
    {
      "calledWith": {
        "accessToken": "test-token",
        "username": "d",
      },
      "name": "set_access_token",
    },
  ],
  [
    "emits",
    {
      "calledWith": {
        "data": {
          "is_logged_in": true,
        },
        "isFrontendEmit": true,
      },
      "name": "auth",
    },
  ],
  [
    "invoke",
    {
      "calledWith": {
        "notes": [
          {
            "content": {
              "body": "😬ö",
              "delta": {
                "ops": [
                  {
                    "insert": "Note with special characters

😬ö
  sss
	
Ketchup
Salt 
Sugar

¯\\_(ツ)_/¯

──────────────────────────────────
──────▄▀▀▄────────────────▄▀▀▄────
─────▐▒▒▒▒▌──────────────▌▒▒▒▒▌───
─────▌▒▒▒▒▐─────────────▐▒▒▒▒▒▐───
────▐▒▒▒▒▒▒▌─▄▄▄▀▀▀▀▄▄▄─▌▒▒▒▒▒▒▌──
───▄▌▒▒▒▒▒▒▒▀▒▒▒▒▒▒▒▒▒▒▀▒▒▒▒▒▒▐───
─▄▀▒▐▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌───
▐▒▒▒▌▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▐───
▌▒▒▌▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌──
▒▒▐▒▒▒▒▒▒▒▒▒▄▀▀▀▀▄▒▒▒▒▒▄▀▀▀▀▄▒▒▐──
▒▒▌▒▒▒▒▒▒▒▒▐▌─▄▄─▐▌▒▒▒▐▌─▄▄─▐▌▒▒▌─
▒▐▒▒▒▒▒▒▒▒▒▐▌▐█▄▌▐▌▒▒▒▐▌▐█▄▌▐▌▒▒▐─
▒▌▒▒▒▒▒▒▒▒▒▐▌─▀▀─▐▌▒▒▒▐▌─▀▀─▐▌▒▒▒▌
▒▌▒▒▒▒▒▒▒▒▒▒▀▄▄▄▄▀▒▒▒▒▒▀▄▄▄▄▀▒▒▒▒▐
▒▌▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▄▄▄▒▒▒▒▒▒▒▒▒▒▒▐
▒▌▒▒▒▒▒▒▒▒▒▒▒▒▀▒▀▒▒▒▀▒▒▒▀▒▀▒▒▒▒▒▒▐
▒▌▒▒▒▒▒▒▒▒▒▒▒▒▒▀▒▒▒▄▀▄▒▒▒▀▒▒▒▒▒▒▒▐
▒▐▒▒▒▒▒▒▒▒▒▒▀▄▒▒▒▄▀▒▒▒▀▄▒▒▒▄▀▒▒▒▒▐
▒▓▌▒▒▒▒▒▒▒▒▒▒▒▀▀▀▒▒▒▒▒▒▒▀▀▀▒▒▒▒▒▒▐
▒▓▓▌▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌
▒▒▓▐▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌─
▒▒▓▓▀▀▄▄▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▐──
▒▒▒▓▓▓▓▓▀▀▄▄▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▄▄▀▀▒▌─
▒▒▒▒▒▓▓▓▓▓▓▓▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▒▒▒▒▒▐─
▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌
▒▒▒▒▒▒▒█▒█▒█▀▒█▀█▒█▒▒▒█▒█▒█▒▒▒▒▒▒▐
▒▒▒▒▒▒▒█▀█▒█▀▒█▄█▒▀█▒█▀▒▀▀█▒▒▒▒▒▒▐
▒▒▒▒▒▒▒▀▒▀▒▀▀▒▀▒▀▒▒▒▀▒▒▒▀▀▀▒▒▒▒▒▒▐
█▀▄▒█▀▄▒█▀▒█▀█▒▀█▀▒█▒█▒█▒█▄▒█▒▄▀▀▐
█▀▄▒█▀▄▒█▀▒█▄█▒▒█▒▒█▀█▒█▒█▀██▒█▒█▐
▀▀▒▒▀▒▀▒▀▀▒▀▒▀▒▒▀▒▒▀▒▀▒▀▒▀▒▒▀▒▒▀▀▐
▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▐

",
                  },
                ],
              },
              "title": "Note with special characters",
            },
            "id": "id",
            "timestamp": 0,
          },
          {
            "content": {
              "body": "amet consectetur adipisicing",
              "delta": {
                "ops": [
                  {
                    "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
                  },
                ],
              },
              "title": "Amet consectetur adipisicing",
            },
            "id": "id",
            "timestamp": 0,
          },
          {
            "content": {
              "body": "amet consectetur adipisicing",
              "delta": {
                "ops": [
                  {
                    "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
                  },
                ],
              },
              "title": "Amet consectetur adipisicing",
            },
            "id": "id",
            "timestamp": 0,
          },
          {
            "content": {
              "body": "amet consectetur adipisicing",
              "delta": {
                "ops": [
                  {
                    "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
                  },
                ],
              },
              "title": "Amet consectetur adipisicing",
            },
            "id": "id",
            "timestamp": 0,
          },
          {
            "content": {
              "body": "amet consectetur adipisicing",
              "delta": {
                "ops": [
                  {
                    "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
                  },
                ],
              },
              "title": "Amet consectetur adipisicing",
            },
            "id": "id",
            "timestamp": 0,
          },
          {
            "content": {
              "body": "amet consectetur adipisicing",
              "delta": {
                "ops": [
                  {
                    "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
                  },
                ],
              },
              "title": "Amet consectetur adipisicing",
            },
            "id": "id",
            "timestamp": 0,
          },
          {
            "content": {
              "body": "amet consectetur adipisicing",
              "delta": {
                "ops": [
                  {
                    "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
                  },
                ],
              },
              "title": "Amet consectetur adipisicing",
            },
            "id": "id",
            "timestamp": 0,
          },
          {
            "content": {
              "body": "amet consectetur adipisicing",
              "delta": {
                "ops": [
                  {
                    "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
                  },
                ],
              },
              "title": "Amet consectetur adipisicing",
            },
            "id": "id",
            "timestamp": 0,
          },
          {
            "content": {
              "body": "amet consectetur adipisicing",
              "delta": {
                "ops": [
                  {
                    "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
                  },
                ],
              },
              "title": "Amet consectetur adipisicing",
            },
            "id": "id",
            "timestamp": 0,
          },
          {
            "content": {
              "body": "amet consectetur adipisicing",
              "delta": {
                "ops": [
                  {
                    "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
                  },
                ],
              },
              "title": "Amet consectetur adipisicing",
            },
            "id": "id",
            "timestamp": 0,
          },
        ],
      },
      "name": "sync_local_notes",
    },
  ],
  [
    "invoke",
    {
      "calledWith": {
        "username": "d",
      },
      "name": "get_access_token",
    },
  ],
  [
    "request",
    {
      "calledWith": {
        "body": undefined,
        "headers": {
          "Authorization": "Bearer test-token",
          "Content-Security-Policy": "default-src 'self'; connect-src http://localhost:8000;",
          "Content-Type": "application/json",
          "X-Username": "d",
        },
        "method": "POST",
      },
      "name": "/auth/logout",
    },
  ],
  [
    "invoke",
    {
      "calledWith": {
        "username": "d",
      },
      "name": "delete_access_token",
    },
  ],
  [
    "emits",
    {
      "calledWith": {
        "data": {
          "is_logged_in": false,
        },
        "isFrontendEmit": true,
      },
      "name": "auth",
    },
  ],
]
`;

exports[`Logout > Logs out on click > document.body.outerHTML 1`] =
  `"<body class="theme--system"></body>"`;

exports[`Logout > Logs out on click > mockDb 1`] = `
{
  "deletedNoteIds": [],
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`Logout > Logs out on click > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [
    {
      "content": {
        "body": "😬ö",
        "delta": {
          "ops": [
            {
              "insert": "Note with special characters

😬ö
  sss
	
Ketchup
Salt 
Sugar

¯\\_(ツ)_/¯

──────────────────────────────────
──────▄▀▀▄────────────────▄▀▀▄────
─────▐▒▒▒▒▌──────────────▌▒▒▒▒▌───
─────▌▒▒▒▒▐─────────────▐▒▒▒▒▒▐───
────▐▒▒▒▒▒▒▌─▄▄▄▀▀▀▀▄▄▄─▌▒▒▒▒▒▒▌──
───▄▌▒▒▒▒▒▒▒▀▒▒▒▒▒▒▒▒▒▒▀▒▒▒▒▒▒▐───
─▄▀▒▐▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌───
▐▒▒▒▌▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▐───
▌▒▒▌▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌──
▒▒▐▒▒▒▒▒▒▒▒▒▄▀▀▀▀▄▒▒▒▒▒▄▀▀▀▀▄▒▒▐──
▒▒▌▒▒▒▒▒▒▒▒▐▌─▄▄─▐▌▒▒▒▐▌─▄▄─▐▌▒▒▌─
▒▐▒▒▒▒▒▒▒▒▒▐▌▐█▄▌▐▌▒▒▒▐▌▐█▄▌▐▌▒▒▐─
▒▌▒▒▒▒▒▒▒▒▒▐▌─▀▀─▐▌▒▒▒▐▌─▀▀─▐▌▒▒▒▌
▒▌▒▒▒▒▒▒▒▒▒▒▀▄▄▄▄▀▒▒▒▒▒▀▄▄▄▄▀▒▒▒▒▐
▒▌▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▄▄▄▒▒▒▒▒▒▒▒▒▒▒▐
▒▌▒▒▒▒▒▒▒▒▒▒▒▒▀▒▀▒▒▒▀▒▒▒▀▒▀▒▒▒▒▒▒▐
▒▌▒▒▒▒▒▒▒▒▒▒▒▒▒▀▒▒▒▄▀▄▒▒▒▀▒▒▒▒▒▒▒▐
▒▐▒▒▒▒▒▒▒▒▒▒▀▄▒▒▒▄▀▒▒▒▀▄▒▒▒▄▀▒▒▒▒▐
▒▓▌▒▒▒▒▒▒▒▒▒▒▒▀▀▀▒▒▒▒▒▒▒▀▀▀▒▒▒▒▒▒▐
▒▓▓▌▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌
▒▒▓▐▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌─
▒▒▓▓▀▀▄▄▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▐──
▒▒▒▓▓▓▓▓▀▀▄▄▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▄▄▀▀▒▌─
▒▒▒▒▒▓▓▓▓▓▓▓▀▀▀▀▀▀▀▀▀▀▀▀▀▀▀▒▒▒▒▒▐─
▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▌
▒▒▒▒▒▒▒█▒█▒█▀▒█▀█▒█▒▒▒█▒█▒█▒▒▒▒▒▒▐
▒▒▒▒▒▒▒█▀█▒█▀▒█▄█▒▀█▒█▀▒▀▀█▒▒▒▒▒▒▐
▒▒▒▒▒▒▒▀▒▀▒▀▀▒▀▒▀▒▒▒▀▒▒▒▀▀▀▒▒▒▒▒▒▐
█▀▄▒█▀▄▒█▀▒█▀█▒▀█▀▒█▒█▒█▒█▄▒█▒▄▀▀▐
█▀▄▒█▀▄▒█▀▒█▄█▒▒█▒▒█▀█▒█▒█▀██▒█▒█▐
▀▀▒▒▀▒▀▒▀▀▒▀▒▀▒▒▀▒▒▀▒▀▒▀▒▀▒▒▀▒▒▀▀▐
▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▐

",
            },
          ],
        },
        "title": "Note with special characters",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
    {
      "content": {
        "body": "amet consectetur adipisicing",
        "delta": {
          "ops": [
            {
              "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
            },
          ],
        },
        "title": "Amet consectetur adipisicing",
      },
      "id": "id",
      "timestamp": 0,
    },
  ],
  "selectedNote": {
    "content": {
      "body": "amet consectetur adipisicing",
      "delta": {
        "ops": [
          {
            "insert": "Amet consectetur adipisicing\\n\\namet consectetur adipisicing\\namet consectetur adipisicing\\namet consectetur adipisicing\\n",
          },
        ],
      },
      "title": "Amet consectetur adipisicing",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Logout > Logs out on click > openedPopup 1`] = `undefined`;

exports[`Logout > Logs out on click > selectedTheme 1`] = `"System"`;

exports[`Logout > Logs out on click > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`Logout > Logs out on click > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Logout > Mounts > KeyStore.getKey() 1`] = `undefined`;

exports[`Logout > Mounts > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Logout > Mounts > Storage.get('THEME') 1`] = `null`;

exports[`Logout > Mounts > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Logout > Mounts > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Logout > Mounts > Storage.get('USERNAME') 1`] = `null`;

exports[`Logout > Mounts > allCalls 1`] = `[]`;

exports[`Logout > Mounts > document.body.outerHTML 1`] =
  `"<body class="theme--system"></body>"`;

exports[`Logout > Mounts > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`Logout > Mounts > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Logout > Mounts > openedPopup 1`] = `undefined`;

exports[`Logout > Mounts > selectedTheme 1`] = `"System"`;

exports[`Logout > Mounts > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "set": [Function],
    "size": 0,
    "store": [Function],
  },
  "username": "",
}
`;

exports[`Logout > Mounts > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;
